### PR TITLE
[keymap] Keychron Q2 ANSI encoder with Caps Lock, Fn Layer and Caps Word RGB indicators 

### PR DESCRIPTION
**Pull request by mkillewald**
Adding keymap for Keychron Q2 ANSI encoder with Caps Lock, Fn Layer and Caps Word RGB indicators 

## Description
 
- Retains Keychron factory keycodes for Mission Control, Launchpad, Siri, Cortana, Task View, File Explorer, and others. 

- Caps Word enabled with RGB indicator
    - Activated by double tapping left Shift
        - can change activation to tapping both left and right Shift by replacing define DOUBLE_TAP_SHIFT_TURNS_ON_CAPS_WORD with #define BOTH_SHIFTS_TURNS_ON_CAPS_WORD in config.h
    - Left Shift, left Control and left Option (or left Windows key) will light when Caps Word is enabled
        - #define CAPS_WORD_INDICATOR_COLOR [color] in config.h to set the backlight color used for the indicator when Caps Word is enabled (default: red)
        - remove #define CAPS_WORD_LIGHT_LOWER_LEFT_CORNER from config.h if you wish for the indicator to only light left Shift

- Caps Lock RGB indicator
    - the Caps Lock key will light when Caps Lock is enabled with the following options:
        - #define CAPS_LOCK_INDICATOR_COLOR [color] in config.h to set the backlight color used for the indicator when Caps Lock is enabled (default: red)
        - Fn2+Z (keycode: LTTOG) will toggle lighting the TAB key when Caps Lock is enabled. This is useful with non backlit keycaps/legends. (default: off)
        - Fn2+X (keycode: LATOG) will toggle lighting all the alpha keys when Caps Lock is enabled. (default: off)

- Dynamic Fn layer RGB indicator
    - When either Fn1 or Fn2 is held down, any keys defined on the Fn layers in this firmware or in VIA will be highlighted with the following options:
        - #define FN1_LAYER_COLOR [color] in config.h to set a static color for defined keys on Fn1 layer (default: orange)
        - #define FN2_LAYER_COLOR [color] in config.h to set a static color for defined keys on Fn2 layer (default: green)
        - Fn2+C (keycode: TKTOG) will toggle turning off RGB for keys with no definition (default: RGB off)
        - Fn2+V (keycode: FCTOG) will toggle lighting the defined Fn layer keys with the static color set with FN1_LAYER_COLOR or FN2_LAYER_COLOR (default: static color off)

All custom keycodes described above can be moved to different keys in VIA by using the following keycodes:
- CUSTOM(11) = LTTOG (default: Fn2+Z) Caps Lock Light Tab Toggle
- CUSTOM(12) = LATOG (default: Fn2+X) Caps Lock Light Alpha Toggle
- CUSTOM(13) = TKTOG (default: Fn2+C) Fn layer Trasnparent Key Toggle
- CUSTOM(14) = FCTOG (default: Fn2+V) Fn layer Color Toggle
 
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by mkillewald** [comment]
> Not to bother you. My parents just got me the Keychron Q2, I love it so far but I want to add a few things like you have here. I'm new to all of this and I would love to use what you have here, mainly because I always hit the caps lock lol. My question is, how do I use any of this?. I don' know how to download it or use it. Sorry again, new to all of this. Thank you for any help you are willing to give me

This is a pull request to have this code merged into the QMK repository. Once this code has been merged, you would just clone this QMK repository to your PC and compile with:

qmk compile -kb keychron/q2/ansi_encoder -km mkillewald

Then flash the resulting .bin file to your keyboard with:

qmk flash -kb keychron/q2/ansi_encoder -km mkillewald

See the QMK docs for further information: [https://docs.qmk.fm/#/newbs](https://docs.qmk.fm/#/newbs)

File: keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/keychron_q2_ansi_knob_configure_mkillewald.json
Comment (tzarc) [inline keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/keychron_q2_ansi_knob_configure_mkillewald.json:1]: Remove this file.

File: keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/keymap.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+} bootloader_state;
+#endif
+
+enum my_keycodes {
+#ifdef VIA_ENABLE
+    KC_LIGHT_TAB_TOGGLE = QK_KB_11,     // USER11 here gave compile error (not defined)
+                                        // indexed at 11 to appear after Keychron's keycodes
+#else
+    KC_LIGHT_TAB_TOGGLE = SAFE_RANGE,
+#endif
```
Comment (tzarc) [inline keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/keymap.c:42]:
```suggestion
    KC_LIGHT_TAB_TOGGLE = QK_USER_0,
```

**Review by tzarc — CHANGES_REQUESTED** [review]

File: keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/readme.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+- CUSTOM(13) = TKTOG (default: Fn2+C) Fn layer Trasnparent Key Toggle
+- CUSTOM(14) = FCTOG (default: Fn2+V) Fn layer Color Toggle
+
+RGB must be toggled on for all indicators to function. If you do not want an RGB mode active but still want the indicators, toggle RGB on and turn the brightness all the way off. The indicators will remain at full brightness.
+
+Please make sure to save any customizations you have made in VIA to a .json file before flashing the firmware. Sometimes it has been necessary to re-apply those changes in VIA after flashing the firmware. If that is the case, you will most likely need to manually add the custom keycodes back in.
+
+## Optional .json Configuration File
+
+[keychron_q2_ansi_knob_configure_mkillewald.json](https://drive.google.com/file/d/1fCnly9thzn6KfIpd2yOrHi0w9CQgaIxn/view?usp=share_link)
```
Comment (fauxpark) [inline keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/readme.md:38]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/readme.md (hunk 2 above)
Comment (mkillewald) [reply keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/readme.md:38]: Can I ask why the link needs to be removed?  Its just doing some useful keymap layout changes that I didn't want to hardcode into the firmware, but if you'd rather I do hardcode those changes I can do that. Just curious as to the reasoning behind this.

Comment (mkillewald) [reply keyboards/keychron/q2/ansi_encoder/keymaps/mkillewald/readme.md]: The .json file has been removed and layout changes previously made by this file are now baked into the firmware.

**Comment by mkillewald** [comment]
Please, do not merge this PR.  I've created a new PR #20437 which supercedes these changes. Sorry for any confusion.

**Pull request closed without merging**